File: 1382. Balance a Binary Search Tree/Solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func balanceBST(root *TreeNode) *TreeNode {
    nodes := []int{}
    dfs(root,&nodes)
    return build(0,len(nodes)-1, &nodes)
}

func build(start, end int, nodes *[]int) *TreeNode{
    mid := (start + end)/2
    node := TreeNode{
        Val : (*nodes)[mid],
    }
    if mid - 1 >= start {
        node.Left = build(start, mid-1, nodes)
    }
    if mid + 1 <= end {
        node.Right = build(mid+1, end, nodes)
    }
    return &node
}
func dfs(root *TreeNode, nodes *[]int){
    if root == nil {
        return 
    }
    dfs(root.Left, nodes)
    (*nodes) = append((*nodes), root.Val)
    dfs(root.Right, nodes)
}

File: 1382. Balance a Binary Search Tree/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        def dfs(root):
            if not root:
                return 
            dfs(root.left)
            nodes.append(root.val)
            dfs(root.right)
        
        def build(start,end):
            if start > end:
                return None
            mid = (start + end )//2
            return TreeNode(
                val = nodes[mid],
                left = build(start, mid-1), 
                right = build(mid+1,end)
            )
        nodes = []
        dfs(root)
        return build(0,len(nodes)-1)
